### PR TITLE
[Fix](Variant) fix variant lost null info after `cast_column`

### PR DESCRIPTION
**Pull request by eldenmoon**
This could result incorrect output in hirachinal cases

```
sql """insert into ${table_name} values (-3, '{"a" : 1, "b" : 1.5, "c" : [1, 2, 3]}')"""
sql """insert into  ${table_name} select -2, '{"a": 11245, "b" : [123, {"xx" : 1}], "c" : {"c" : 456, "d" : "null", "e" : 7.111}}'  as json_str
            union  all select -1, '{"a": 1123}' as json_str union all select *, '{"a" : 1234, "xxxx" : "kaana"}' as json_str from numbers("number" = "4096") limit 4096 ;"""

mysql> select v["c"] from var_rs where k = -3 or k = -2;
+----------------------+
| element_at(`v`, 'c') |
+----------------------+
| [1,2,3]              |
| []                   |
+----------------------+
2 rows in set (0.04 sec)
```

## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

File: be/src/vec/common/schema_util.cpp [hunk 1]
```diff
@@ -152,13 +152,31 @@ Status cast_column(const ColumnWithTypeAndName& arg, const DataTypePtr& type, Co
     Block tmp_block {arguments};
     size_t result_column = tmp_block.columns();
     auto ctx = FunctionContext::create_context(nullptr, {}, {});
+
+    // To prevent from null info lost, we should not call function since the function framework will wrap
+    // nullable to Variant instead of the root of Variant
+    // Nullable(Array(int)) -> Variant<Nullable(Array(int))>
```
Comment (xiaokang) [inline be/src/vec/common/schema_util.cpp:158]: Is this example right or wrong? Can you write both a right and a wrong example in comment?

File: be/src/vec/common/schema_util.cpp [hunk 2]
```diff
@@ -152,13 +152,31 @@ Status cast_column(const ColumnWithTypeAndName& arg, const DataTypePtr& type, Co
...
+    if (WhichDataType(remove_nullable(type)).is_variant_type()) {
+        // set variant root column/type to from column/type
+        auto variant = ColumnObject::create(true /*always nullable*/);
+        CHECK(arg.column->is_nullable());
```
Comment (xiaokang) [inline be/src/vec/common/schema_util.cpp:162]: Why is src column always nullable?

File: be/src/vec/common/schema_util.cpp (hunk 1 above)
Comment (eldenmoon) [reply be/src/vec/common/schema_util.cpp:158]: right:  Nullable(Array(int)) -> Nullable(Variant<Nullable(Array(int)))
wrong:   Nullable(Array(int)) -> NullableVariant<(Array(int)))

File: be/src/vec/common/schema_util.cpp (hunk 2 above)
Comment (eldenmoon) [reply be/src/vec/common/schema_util.cpp:162]: all the subcolumn is stored ad nullable

File: be/src/vec/common/schema_util.cpp (hunk 1 above)
Comment (eldenmoon) [reply be/src/vec/common/schema_util.cpp]: done

**Comment by eldenmoon** [comment]
run buildall

**Review by xiaokang — APPROVED** [review]
LGTM

**Review by qidaye — APPROVED** [review]
LGTM

**Pull request merged**